src/emscatter.jl

```diff
@@ -0,0 +1,323 @@
+#
+# ```math
+# Y_m(z)^\prime=\frac{Y_{m-1}(z)-Y_{m+1}(z)}{2}
+# ```
+#
+# where $Y_m$ denotes any Bessel functions (Hankel functions).
+#
+# 
+# Finally, the analytical field is ($1/2k$ is the amplitude that comes from the unit line source excitation):
+# ```math
+# H(r>R)=\frac{1}{2k}\sum_m\left[\alpha_mi^mH_m^1(kr)+J_m(kr)\right]e^{im\theta}
+# ```
+#
+# ```math
+# H(r<=R)=\frac{1}{2k}\sum_m\beta_mi^mJ_m(nkr)e^{im\theta}
+# ```
+# 
+
+# ### Define the analytical functions
+using SpecialFunctions
+dbesselj(m,z) = (besselj(m-1,z)-besselj(m+1,z))/2
+dhankelh1(m,z)= (hankelh1(m-1,z)-hankelh1(m+1,z))/2
+α(m,n,z) = (besselj(m,n*z)*dbesselj(m,z)-1/n*besselj(m,z)*dbesselj(m,n*z))/(1/n*hankelh1(m,z)*dbesselj(m,n*z)-besselj(m,n*z)*dhankelh1(m,z))
+β(m,n,z) = (hankelh1(m,z)*dbesselj(m,z)-besselj(m,z)*dhankelh1(m,z))/(1/n*dbesselj(m,n*z)*hankelh1(m,z)-besselj(m,n*z)*dhankelh1(m,z))
+
+function H_t(x,xc,r,ϵ,λ)
+    n = √ϵ
+    k = 2*π/λ
+    θ = angle(x[1]-xc[1]+1im*(x[2]-xc[2]))+π
+    M = 40 # Number of Bessel function basis used
+    H0 = 0
+    if norm([x[1]-xc[1],x[2]-xc[2]])<=r
+        for m=-M:M
+            H0 += β(m,n,k*r)*cis(m*θ)*besselj(m,n*k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    else
+        for m=-M:M
+            H0 += α(m,n,k*r)*cis(m*θ)*hankelh1(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))+cis(m*θ)*besselj(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    end
+    return 1im/(2*k)*H0
+end
+#
+# ### Construct the analytical solution in finite element basis
+# Basically we are solving a equation $u=H_t$ which has a weak form $a_t(u,v)=uv$ and source term $b_t(v)=H_t(x)$
+a_t(u,v) = ∫(u*v)*dΩ
+b_t(v) = ∫(v*(x->H_t(x,xc,r,ϵ1,λ)))*dΩ
+op_t = AffineFEOperator(a_t,b_t,U,V)
+uh_t = solve(op_t)
+
+# ## Output and compare results
+# 
+# The simulated field is shown below. We can see that the simulated fields and the analytical solution matched closed except for the top and PML regions. This is because the simulated source generate plane waves in two directions but we only consider the downward propagating wave in the analytical solution and the PML effect is also not considered in the analytical solution. Therefore, we just need to focus on the "center" regions which excludes the PML and top region above the source, the difference is within 2% of the field amplitude integral. As we increase the resolution, this difference should decrease (until it becomes limited by the PML reflection coefficient from $\sigma_0$, the number of Bessel function basis $M$ or by floating-point error.)
+# ![](../assets/emscatter/Results.png)
+
+# ### Save to file and view
+writevtk(Ω,"demo",cellfields=["Real"=>real(uh),
+        "Imag"=>imag(uh),
+        "Norm"=>abs2(uh),
+        "Real_t"=>real(uh_t),
+        "Imag_t"=>imag(uh_t),
+        "Norm_t"=>abs2(uh_t),
+        "Difference"=>abs2(uh_t-uh)])
+# ### Compare the difference in the "center" region
+function AnalyticalBox(x) # Get the "center" region
```

Comment:
If the center region is named with a tag in the mesh, you don't need to introduce this function. It is more eficient and accurate to just integrate in the selected region. E.g.
```julia
 Ωc = Triangulation(model,tags="Cylinder")
 dΩc = Measure(Ωc,2*order)
```
then integrate using `dΩc`